contracts/integrations/contracts/dydx/ZeroExV3ExchangeWrapper.sol

```diff
@@ -0,0 +1,331 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pragma solidity 0.5.9;
+pragma experimental ABIEncoderV2;
+
+import { SafeMath } from "./libs/SafeMath.sol";
+import { IExchange } from "@0x/contracts-exchange/contracts/src/interfaces/IExchange.sol";
+import { LibFillResults } from "@0x/contracts-exchange-libs/contracts/src/LibFillResults.sol";
+import { LibOrder } from "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+import { ExchangeReader } from "./interfaces/ExchangeReader.sol";
+import { ExchangeWrapper } from "./interfaces/ExchangeWrapper.sol";
+import { MathHelpers } from "./libs/MathHelpers.sol";
+import { TokenInteract } from "./libs/TokenInteract.sol";
+
+
+/**
+ * @title ZeroExV3ExchangeWrapper
+ * @author dYdX/0x
+ *
+ * dYdX ExchangeWrapper to interface with 0x Version 3
+ */
+contract ZeroExV3ExchangeWrapper is
+    ExchangeWrapper,
+    ExchangeReader
+{
+    using SafeMath for uint256;
+    using TokenInteract for address;
+
+    // ============ State Variables ============
+
+    // msg.senders that will put the correct tradeOriginator in callerData when doing an exchange
+    mapping (address => bool) public TRUSTED_MSG_SENDER;
+
+    // address of the ZeroEx V3 Exchange
+    address public ZERO_EX_EXCHANGE;
+
+    // address of the ZeroEx ERC20Proxy
+    address public ZERO_EX_TOKEN_PROXY;
+
+    // address of the WETH token contract
+    address public WETH_TOKEN;
+
+    // ============ Constructor ============
+
+    constructor(
+        address zeroExExchange,
+        address zeroExProxy,
+        address weth,
+        address[] memory trustedMsgSenders
+    )
+        public
+    {
+        ZERO_EX_EXCHANGE = zeroExExchange;
+        ZERO_EX_TOKEN_PROXY = zeroExProxy;
+        WETH_TOKEN = weth;
+
+        for (uint256 i = 0; i < trustedMsgSenders.length; i++) {
+            TRUSTED_MSG_SENDER[trustedMsgSenders[i]] = true;
+        }
+    }
+
+    // ============ Public Functions ============
+
+    function exchange(
+        address tradeOriginator,
+        address receiver,
+        address makerToken,
+        address takerToken,
+        uint256 requestedFillAmount,
+        bytes calldata orderData
+    )
+        external
+        returns (uint256)
+    {
+        // prepare the exchange
+        (LibOrder.Order memory order, address takerFeeToken) = parseOrder(orderData, makerToken, takerToken);
+        bytes memory signature = parseSignature(orderData);
+
+        // transfer taker fee from trader if applicable
+        transferTakerFee(
+            order,
+            tradeOriginator,
+            requestedFillAmount,
+            takerFeeToken
+        );
+
+        // transfer protocol fee from trader if applicable
+        transferProtocolFee(tradeOriginator);
```

Comment:
I should also note that the other option is for this contract always hold a WETH balance of its own.
The benefit of such an approach would be that the `tradeOriginator` wouldn't need to approve this contract to transfer WETH on their behalf.
The downsides would be that dYdX would need to periodically top up this contract's WETH balance, and might want to adjust prices to account for the fact that they are paying the protocol fee on the user's behalf (this would probably be easy if they're already doing this for gas)